lib/sensors/GPS.cpp

```diff
@@ -0,0 +1,483 @@
+//
+// Created by gratt on 2020-04-30.
+//
+
+#include "GPS.h"
+GPS::GPS(PinName tx, PinName rx, uint32_t baud_rate) : m_uart(tx, rx, DEFAULT_BAUD_RATE) {
+	m_rollingChecksumA = 0;
+	m_rollingChecksumB = 0;
+	m_ubxFrameCounter = 0;
+	m_ignoreThisPayload = false;
+	m_ack_flag = false;
+	m_nack_flag = false;
+	//TODO: verify that ACKs get read no matter what target class/id is selected
+	m_target_class = UBX_CLASS_NAV; //default class target
+	m_target_msgID = UBX_NAV_PVT; //default id target
+	setSerialRate(baud_rate, COM_PORT_UART1);
+	m_uart.baud(baud_rate);
+	wait_ms(50); //ACK gets lost during baud rate change, so no way to verify here
+	m_uart.attach(callback(this, &GPS::onByteReceive));
+}
+
+void GPS::configure() {
+	//TODO: Determine GNSS constellation configuration (necessary? best config?)
+	//allow uart1 to output UBX messages
+	setPortOutput(COM_PORT_UART1, COM_TYPE_UBX);
+	if (!waitForAck()) {
+		//TODO: handle NACK response
+	}
+	setNavigationFrequency(NAV_FREQUENCY);
+	if (!waitForAck()) {
+		//TODO: handle NACK response
+	}
+	//configure UBX_NAV_PVT to output from uart 1 every navigation cycle
+	configureMessage(UBX_CLASS_NAV, UBX_NAV_PVT, COM_PORT_UART1, 1);
+	if (!waitForAck()) {
+		//TODO: handle NACK response
+	}
+	//turn default NEMA messages off
+	for (const uint8_t message : DEFAULT_ENABLED_NMEA) {
+		configureMessage(UBX_CLASS_NMEA, message, COM_PORT_UART1, 0); //0 sendRate message off
+		if (!waitForAck()) {
+			//TODO: handle NACK response
+		}
+	}
+
+}
+
+bool GPS::waitForAck() {
+	//TODO: need error checking for failed checksums, etc
+	while (!m_ack_flag) { //wait for ack
+		if (m_nack_flag) {
+			m_nack_flag = false;
+			return false;
+		}
+	}
+	m_ack_flag = false;
+	return true;
+}
+
+void GPS::onByteReceive() {
+	process(m_uart.getc(), (m_activePacketBuffer == UBLOX_PACKET_PACKETCFG ? &m_packetCfg :
+							m_activePacketBuffer == UBLOX_PACKET_PACKETACK ? &m_packetAck : &m_packetBuf),
```

Comment:
nested ternary operator is not very readable, but it might just be a personal preference